src/ruby/puppetserver-lib/puppet/server/compiler.rb

```diff
@@ -29,6 +29,37 @@ def compile(request_data)
         end
       end
 
+      # Make the node object to be used in compilation. This requests it from
+      # the node indirection and merges local facts and other data.
+      # @api private
+      def create_node(request_data)
+        facts, trusted_facts = process_facts(request_data)
+        certname = request_data['certname']
+        environment = request_data['environment'] || 'production'
+        transaction_uuid = request_data['transaction_uuid']
+        prefer_requested_environment =
+          request_data.dig('options', 'prefer_requested_environment')
+
+        node = Puppet.override(trusted_information: trusted_facts) do
+          Puppet::Node.indirection.find(certname,
+                                        environment: environment,
+                                        facts: facts,
+                                        transaction_uuid: transaction_uuid)
+        end
+
+        if prefer_requested_environment
+          node.environment = environment
+        end
+
+        # Merges facts into the node parameters.
+        # Ensures that facts will be surfaced as top-scope variables,
+        # along with other node parameters.
+        node.merge(facts.values)
```

Comment:
I think all termini do this already, by calling `node.fact_merge` which also sanitizes facts (which we're doing separately as well)

Reply:
Should we get rid of it here, then?